nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py

```diff
@@ -0,0 +1,521 @@
+from nemo.collections.nlp.parts.nlp_overrides import GradScaler
+from nemo.core.optim import MainParamsOptimizerWrapper, prepare_lr_scheduler
+from nemo.utils import AppState, logging
+
+try:
+    from apex.transformer.enums import ModelType
+    from apex.transformer import parallel_state
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    ModelType = ApexGuardDefaults()
+    HAVE_APEX = False
+
+
+__all__ = ["MegatronRetrievalModel"]
+
+
+class MegatronRetrievalModel(MegatronBaseModel):
+    """
+    Megatron Retrieval enhanced language model
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer=trainer)
+
+        # build tokenizer (defaults to nemo supported tokenizers)
+        self._build_tokenizer()
+
+        # manipulate vocabulary (e.g., pad vocabulary for better efficiency)
+        self._build_vocab()
+
+        # TODO does not support PP yet
+        self.model = self.model_provider_func(True, True, True, True)
+
+        self.setup_optimizer_param_groups()
+
+        self.megatron_amp_o2 = cfg.get('megatron_amp_O2', False)
+
+        if self.megatron_amp_o2:
+
+            # Pre-allocate the model on GPU to have master parameters allocated on the same device with matching data type
+            self.model.cuda(torch.cuda.current_device())
+
+            # Model wrapper to convert both model and inputs to half precision
+            self.model = Float16Module(module=self.model, precision=cfg.precision)
+
+        if self.cfg.precision == 32:
+            self.autocast_dtype = torch.float
+        elif self.cfg.precision == 16:
+            self.autocast_dtype = torch.half
+        elif self.cfg.precision == 'bf16':
+            self.autocast_dtype = torch.bfloat16
+        else:
+            raise ValueError('precision must be in [32, 16, "bf16"]')
+
+        self.model.model_type = ModelType.encoder_and_decoder
+        self._reduced_loss_buffer = []
+
+    def _build_tokenizer(self):
+        """
+        Default tokenizer is based on available nemo tokenizers.
+        Override this method to use an external tokenizer.
+        All tokenizers are expected to provide compatible interface.
+        Override default Encoder-decoder tokenizer to use legacy=True for sentencepiece.
+        """
+        self.tokenizer = get_nmt_tokenizer(
+            library=self._cfg.tokenizer.library,
+            model_name=self._cfg.tokenizer.type,
+            tokenizer_model=self.register_artifact("tokenizer.model", self._cfg.tokenizer.model),
+            vocab_file=self.register_artifact("tokenizer.vocab_file", self._cfg.tokenizer.vocab_file),
+            merges_file=self.register_artifact("tokenizer.merge_file", self._cfg.tokenizer.merge_file),
+            legacy=True if self._cfg.tokenizer.library == 'sentencepiece' else False,
+        )
+        # add pad special token
+        if not self.tokenizer.pad_id:
```

Comment:
This looks a little dangerous. What if pad_id = 0? If not 0 will evaluate to True. This will add another pad_id no? Can we make the check more explicit using hasattr(self.tokenizer, "pad_id") or self.tokenizer.pad > 0 or something like that

Reply:
this is how I fix it.
```python
        # add pad special token
        if not hasattr(self.tokenizer, "pad_id"):
            self.tokenizer.add_special_tokens({'pad_token': '<pad>'})
        elif hasattr(self.tokenizer, "pad_id") and (self.tokenizer.pad_id is None or self.tokenizer.pad_id < 0):
            self.tokenizer.add_special_tokens({'pad_token': '<pad>'})

```